saleor/dashboard/order/forms.py

```diff
@@ -362,11 +362,12 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
     def cancel_line(self):
-        if self.line.variant and self.line.variant.track_inventory:
-            deallocate_stock(self.line.variant, self.line.quantity)
-        order = self.line.order
-        self.line.delete()
-        recalculate_order(order)
+        if self.line.variant:
+            if self.line.variant.track_inventory:
+                deallocate_stock(self.line.variant, self.line.quantity)
+            order = self.line.order
```

Comment:
How does canceling order lines work, when there is no variant then? Shouldn't we trigger `delete_order_line` and `recalculate_order` always?

Reply:
I assumed that you can not cancel that line. If we allow that, we could not calculate new `order.weight` (due to lack of variant weight). Calling `delete_order_line` in that case will raise an exception.
Maybe we should have an ability to cancel **all** lines related with non-existing product variants? Then we could calculate `order.weight`.

Reply:
So if we had this case:
- as an admin, I've created a draft order with two lines,
- later I've removed a product variant that was related to one of those lines,
- since I cannot finalize my draft order now, I need to remove the line related to the non-existing variant

It feels totally valid to be able to cancel the line. After removing it, you'd calculate new order weight by iterating over existing lines and summing their variant weights.

Reply:
You're right, but let's consider that case:

- I've created a draft order with three lines,
- I've removed two different product variant that was related to different two lines,
- I would like to remove one of that line (call it *line A*)

I can not calculate weight by iterating over existing lines, because order still contains line without variant. I can not calculate weight by subtraction, because I have no variant related with *line A*.

I see one solution now. Let's assume that invalid draft order (I mean: *draft order with lines related to non-existing variants*) **could** have wrong weight. I will calculate order weight by iterating existing variants - in your case it results proper weight, in my case it results wrong weight.
But I don't like that solution. ;)

Reply:
If two lines are broken you would have to remove both of them anyway and eventually you'll end up with the weight equal the weight of existing variants.

Reply:
Because of that I also think that we could always calculate weight by iterating over lines instead of doing substractions.